overlord/devicestate/devicestate.go

```diff
@@ -202,11 +204,51 @@ func checkGadgetOrKernel(st *state.State, snapInfo, curInfo *snap.Info, _ snap.C
 	return nil
 }
 
+var (
+	gadgetIsCompatible = func(current, new *gadget.Info) error { return nil }
+)
+
+func checkGadgetRemodelCompatible(st *state.State, snapInfo, curInfo *snap.Info, snapf snap.Container, flags snapstate.Flags, deviceCtx snapstate.DeviceContext) error {
+	if release.OnClassic {
+		return nil
+	}
+	if curInfo == nil || curInfo.GetType() != snap.TypeGadget {
+		// only interested in snaps that are currently installed and are
+		// gadget snaps
+		return nil
+	}
+	if deviceCtx == nil || !deviceCtx.ForRemodeling() {
+		// in a remodeling scenario
+		return nil
+	}
+
+	newGadgetYaml, err := snapf.ReadFile("meta/gadget.yaml")
+	if err != nil {
+		return fmt.Errorf("cannot read new gadget metadata: %v", err)
+	}
+
+	currentData, err := gadgetDataFromInfo(curInfo)
+	if err != nil {
+		return fmt.Errorf("cannot read current gadget metadata: %v", err)
+	}
+
+	pendingInfo, err := gadget.InfoFromGadgetYaml(newGadgetYaml, coreGadgetConstraints)
+	if err != nil {
+		return fmt.Errorf("cannot load new gadget metadata: %v", err)
+	}
+
+	if err := gadgetIsCompatible(currentData.Info, pendingInfo); err != nil {
```

Comment:
Is `currentData.Info` the same as `curInfo`? In other words is reading `currentData` really needed?

Reply:
curInfo is *snap.Info, currentData.Info is  `*gadget.Info`